File: source/S2AP/Addresses.cs
using System.Collections.Generic;

namespace S2AP
{
    public static class Addresses
    {
        public const uint TotalTalismanAddress = 0x00067108;
        public const uint TotalOrbAddress = 0x0006702c;
        public const uint TalismanStartAddress = 0x0006b264; // Starts with Summer Forest!
        public const uint OrbStartAddress = 0x00069fd0; // Includes bytes for levels with no orbs!
        public const uint CurrentLevelAddress = 0x00066f54;
        public const uint IsInDemoMode = 0x00067024; // 0x00067044 seems to behave similarly but is 0 on the Demo mode guidebook screen.
        public const uint GameStatus = 0x000681c8;
        // The values at this and the following 3 bytes seem to be 0 only on reset.
        public const uint ResetCheckAddress = 0x00066f14;
        public const uint PlayerLives = 0x0006712c;
        public const uint PlayerHealth = 0x0006a248;

        public const uint TotalGemAddress = 0x000670cc; // 0x00067660 is the HUD display; may need to be edited if this value is modified.
        public const uint LevelGemsAddress = 0x0006ac04; // One entire word per level, including boss levels.

        public const uint SkelosCactiSkillPoint = 0x0006470c;
        public const uint HurricosWindmillSkillPoint = 0x0006470d;
        public const uint ColossusHockeySkillPoint = 0x0006470e;
        public const uint FractureSuperchargeSkillPoint = 0x0006470f;
        public const uint CrushPerfectSkillPoint = 0x00064711;
        public const uint GulpPerfectSkillPoint = 0x00064712;
        public const uint RiptoPerfectSkillPoint = 0x00064713;
        public const uint ScorchTreesSkillPoint = 0x00064714;
        public const uint OceanTimeAttackSkillPoint = 0x00064715;
        public const uint MetroTimeAttackSkillPoint = 0x00064716;
        public const uint IcyTimeAttackSkillPoint = 0x00064717;
        public const uint CanyonTimeAttackSkillPoint = 0x00064718;
        public const uint IdolTikiSkillPoint = 0x00064719;
        public const uint AquariaSeaweedSkillPoint = 0x0006471a;
        public const uint GulpRiptoSkillPoint = 0x0006471b;
        public const uint SkelosCatbatSkillPoint = 0x0006471c;

        public const uint TokenAddress = 0x000646be; // Each token gets 1 word, but a value of 1 on this byte determines if it's collected.

        public const uint SpyroStateAddress = 0x0006a040; // One byte, set to 31 in decimal to kill Spyro for Death Link.

        public const uint BigHeadMode = 0x0698be;
        public const uint FlatSpyroMode = 0x0698bf;
        public const uint SpyroWidth = 0x0698c1;
        public const uint SpyroHeight = 0x0698c5;
        public const uint SpyroLength = 0x0698c9;
        public const uint SpyroColorAddress = 0x0698cc;

        public const uint PermanentFireballAddress = 0x000698bb;
        public const uint DoubleJumpAddress1 = 0x00035ba8;
        public const uint DoubleJumpAddress2 = 0x00035bb8;
        public const uint InvisibleAddress1 = 0x0004c584;
        public const uint InvisibleAddress2 = 0x0004c586;
        public const uint DestructiveSpyroAddress = 0x0006a12a;

        public const uint CrystalBridgeUnlock = 0x00064670;
        public const uint AquariaSubUnlock = 0x00064674;
        public const uint MagmaElevatorUnlock = 0x00064678;
        public const uint GlimmerBridgeUnlock = 0x0006467c;
        public const uint SwimUnlock = 0x00064680;
        public const uint ClimbUnlock = 0x00064684;
        public const uint HeadbashUnlock = 0x00064688;
        public const uint WallToAquariaUnlock = 0x0006468c;
        public const uint ZephyrPortalUnlock = 0x00064698;
        public const uint ShadyPortalUnlock = 0x0006469c;
        public const uint IcyPortalUnlock = 0x000646a0;
        public const uint CanyonPortalUnlock = 0x000646b8;

        public const uint SummerEloraStartText = 0x001818cb;
        public const uint SummerEloraEndText = 0x00181973;
        public const uint SummerEloraWarpStartText = 0x00181a70;
        public const uint SummerEloraWarpEndText = 0x00181abb;

        public const uint AutumnEloraStartText = 0x00191813;
        public const uint AutumnEloraEndText = 0x001918b8;
        public const uint AutumnEloraWarpStartText = 0x00191a84;
        public const uint AutumnEloraWarpEndText = 0x00191acf;

        public static readonly List<uint> SummerLifeBottle1Address = [0x0006ac8f, 7];
        public static readonly List<uint> SummerLifeBottle2Address = [0x0006ac90, 0];
        public static readonly List<uint> SummerLifeBottle3Address = [0x0006ac8b, 6];
        public static readonly List<uint> IdolLifeBottleAddress = [0x0006acd3, 1];
        public static readonly List<uint> ColossusLifeBottleAddress = [0x0006acfc, 6];
        public static readonly List<uint> HurricosLifeBottleAddress = [0x0006ad0f, 5];
        public static readonly List<uint> AquariaLifeBottleAddress = [0x0006ad2f, 5];
        public static readonly List<uint> AutumnLifeBottleAddress = [0x0006adb3, 6];
        public static readonly List<uint> SkelosLifeBottleAddress = [0x0006ade0, 3];
        public static readonly List<uint> CrystalLifeBottleAddress = [0x0006adff, 7];
        public static readonly List<uint> BreezeLifeBottle1Address = [0x0006ae21, 7];
        public static readonly List<uint> BreezeLifeBottle2Address = [0x0006ae22, 4];
        public static readonly List<uint> ZephyrLifeBottleAddress = [0x0006ae3c, 5];
        public static readonly List<uint> ScorchLifeBottleAddress = [0x0006ae78, 2];
        public static readonly List<uint> ShadyLifeBottleAddress = [0x0006ae93, 5];
        public static readonly List<uint> MagmaLifeBottle1Address = [0x0006aeb9, 0];
        public static readonly List<uint> MagmaLifeBottle2Address = [0x0006aeba, 0];
        public static readonly List<uint> MagmaLifeBottle3Address = [0x0006aeba, 1];
        public static readonly List<uint> MagmaLifeBottle4Address = [0x0006aeba, 2];
        public static readonly List<uint> FractureLifeBottleAddress = [0x0006aee1, 0];
        public static readonly List<uint> MysticLifeBottle1Address = [0x0006af61, 4];
        public static readonly List<uint> MysticLifeBottle2Address = [0x0006af61, 5];
        public static readonly List<uint> CloudLifeBottleAddress = [0x0006af72, 1];

        public const uint SummerGemMask = 0x0006ac84;
        public const uint GlimmerGemMask = 0x0006acaa;
        public const uint IdolGemMask = 0x0006acc4;
        public const uint ColossusGemMask = 0x0006ace7;
        public const uint HurricosGemMask = 0x0006ad04;
        public const uint AquariaGemMask = 0x0006ad24;
        public const uint SunnyGemMask = 0x0006ad4d;
        public const uint AutumnGemMask = 0x0006ada7;
        public const uint SkelosGemMask = 0x0006adcc;
        public const uint CrystalGemMask = 0x0006adf1;
        public const uint BreezeGemMask = 0x0006ae14;
        public const uint ZephyrGemMask = 0x0006ae2b;
        public const uint ScorchGemMask = 0x0006ae6c;
        public const uint ShadyGemMask = 0x0006ae88;
        public const uint MagmaGemMask = 0x0006aeaa;
        public const uint FractureGemMask = 0x0006aecf;
        public const uint WinterGemMask = 0x0006af25;
        public const uint MysticGemMask = 0x0006af4d;
        public const uint CloudGemMask = 0x0006af6e;
        public const uint RoboticaGemMask = 0x0006afac;
        public const uint MetropolisGemMask = 0x0006afca;
    }
}

File: apworld/spyro2/Addresses.py
class RAM:
    TotalTalismanAddress = 0x00067108
    TotalOrbAddress = 0x0006702c
    TalismanStartAddress = 0x0006b264 # Starts with Summer Forest!
    OrbStartAddress = 0x00069fd0 # Includes bytes for levels with no orbs!
    CurrentLevelAddress = 0x00066f54
    IsInDemoMode = 0x00067024 # 0x00067044 seems to behave similarly but is 0 on the Demo mode guidebook screen.
    GameStatus = 0x000681c8
    # The values at this and the following 3 bytes seem to be 0 only on reset.
    ResetCheckAddress = 0x00066f14
    PlayerLives = 0x0006712c
    PlayerHealth = 0x0006a248
    SpiritParticlesAddress = 0x00066fb0

    PlayerZPos = 0x69FF8
    PlayerAnimationLength = 0x6A04C
    PlayerVelocityStatus = 0x6A044

    RiptoDoorOrbRequirementAddress = 0x0007695c
    RiptoDoorOrbDisplayAddress = 0x000769e0

    TotalGemAddress = 0x000670cc # 0x00067660 is the HUD display; may need to be edited if this value is modified.
    LevelGemsAddress = 0x0006ac04 # One entire word per level, including boss levels.

    SkelosCactiSkillPoint = 0x0006470c
    HurricosWindmillSkillPoint = 0x0006470d
    ColossusHockeySkillPoint = 0x0006470e
    FractureSuperchargeSkillPoint = 0x0006470f
    CrushPerfectSkillPoint = 0x00064711
    GulpPerfectSkillPoint = 0x00064712
    RiptoPerfectSkillPoint = 0x00064713
    ScorchTreesSkillPoint = 0x00064714
    OceanTimeAttackSkillPoint = 0x00064715
    MetroTimeAttackSkillPoint = 0x00064716
    IcyTimeAttackSkillPoint = 0x00064717
    CanyonTimeAttackSkillPoint = 0x00064718
    IdolTikiSkillPoint = 0x00064719
    AquariaSeaweedSkillPoint = 0x0006471a
    GulpRiptoSkillPoint = 0x0006471b
    SkelosCatbatSkillPoint = 0x0006471c

    TokenAddress = 0x000646be # Each token gets 1 word, but a value of 1 on this byte determines if it's collected.
    SpyroStateAddress = 0x0006a040 # One byte

    # Cosmetics
    BigHeadMode = 0x0698be
    FlatSpyroMode = 0x0698bf
    SpyroWidth = 0x0698c1
    SpyroHeight = 0x0698c5
    SpyroLength = 0x0698c9
    SpyroColorAddress = 0x0698cc
    PortalTextRed = 0x00064468
    PortalTextGreen = 0x00064469
    PortalTextBlue = 0x0006446a

    RedGemShadow = 0x00064450
    RedGemColor = 0x00064454
    GreenGemShadow = 0x00064460
    GreenGemColor = 0x00064464
    BlueGemShadow = 0x00064470
    BlueGemColor = 0x00064474
    GoldGemShadow = 0x00064480
    GoldGemColor = 0x00064484
    PinkGemShadow = 0x00064490
    PinkGemColor = 0x00064494

    # Abilities
    PermanentFireballAddress = 0x000698bb
    DoubleJumpAddress1 = 0x00035ba8
    DoubleJumpAddress2 = 0x00035bb8
    InvisibleAddress1 = 0x0004c584
    InvisibleAddress2 = 0x0004c586
    DestructiveSpyroAddress = 0x0006a12a

    # Moneybags
    CrystalBridgeUnlock = 0x00064670
    AquariaSubUnlock = 0x00064674
    MagmaElevatorUnlock = 0x00064678
    GlimmerBridgeUnlock = 0x0006467c
    SwimUnlock = 0x00064680
    ClimbUnlock = 0x00064684
    HeadbashUnlock = 0x00064688
    WallToAquariaUnlock = 0x0006468c
    ZephyrPortalUnlock = 0x00064698
    ShadyPortalUnlock = 0x0006469c
    IcyPortalUnlock = 0x000646a0
    CanyonPortalUnlock = 0x000646b8
    
    # Elora text addresses, for giving talisman counts.
    SummerEloraStartText = 0x001818cb
    SummerEloraEndText = 0x00181973
    SummerEloraWarpStartText = 0x00181a70
    SummerEloraWarpEndText = 0x00181abb
    AutumnEloraStartText = 0x00191813
    AutumnEloraEndText = 0x001918b8
    AutumnEloraWarpStartText = 0x00191a84
    AutumnEloraWarpEndText = 0x00191acf

    # Life Bottle Addresses
    SummerLifeBottle1Address = [0x0006ac8f, 7]
    SummerLifeBottle2Address = [0x0006ac90, 0]
    SummerLifeBottle3Address = [0x0006ac8b, 6]
    IdolLifeBottleAddress = [0x0006acd3, 1]
    ColossusLifeBottleAddress = [0x0006acfc, 6]
    HurricosLifeBottleAddress = [0x0006ad0f, 5]
    AquariaLifeBottleAddress = [0x0006ad2f, 5]
    AutumnLifeBottleAddress = [0x0006adb3, 6]
    SkelosLifeBottleAddress = [0x0006ade0, 3]
    CrystalLifeBottleAddress = [0x0006adff, 7]
    BreezeLifeBottle1Address = [0x0006ae21, 7]
    BreezeLifeBottle2Address = [0x0006ae22, 4]
    ZephyrLifeBottleAddress = [0x0006ae3c, 5]
    ScorchLifeBottleAddress = [0x0006ae78, 2]
    ShadyLifeBottleAddress = [0x0006ae93, 5]
    MagmaLifeBottle1Address = [0x0006aeb9, 0]
    MagmaLifeBottle2Address = [0x0006aeba, 0]
    MagmaLifeBottle3Address = [0x0006aeba, 1]
    MagmaLifeBottle4Address = [0x0006aeba, 2]
    FractureLifeBottleAddress = [0x0006aee1, 0]
    MysticLifeBottle1Address = [0x0006af61, 4]
    MysticLifeBottle2Address = [0x0006af61, 5]
    CloudLifeBottleAddress = [0x0006af72, 1]

    # Bit masks for gems (and other flags) in each level.
    SummerGemMask = 0x0006ac84
    GlimmerGemMask = 0x0006acaa
    IdolGemMask = 0x0006acc4
    ColossusGemMask = 0x0006ace7
    HurricosGemMask = 0x0006ad04
    AquariaGemMask = 0x0006ad24
    SunnyGemMask = 0x0006ad4d
    AutumnGemMask = 0x0006ada7
    SkelosGemMask = 0x0006adcc
    CrystalGemMask = 0x0006adf1
    BreezeGemMask = 0x0006ae14
    ZephyrGemMask = 0x0006ae2b
    ScorchGemMask = 0x0006ae6c
    ShadyGemMask = 0x0006ae88
    MagmaGemMask = 0x0006aeaa
    FractureGemMask = 0x0006aecf
    WinterGemMask = 0x0006af25
    MysticGemMask = 0x0006af4d
    CloudGemMask = 0x0006af6e
    RoboticaGemMask = 0x0006afac
    MetropolisGemMask = 0x0006afca

    # Gemsanity code change addresses.
    localGemIncrementAddress = 0x000396C0
    globalGemIncrementAddress = 0x000396d4
    globalGemRespawnFixAddress = 0x0001d36c
    localGemRespawnFixAddress = 0x0001d380
    localGemLoadFixAddress = 0x00076B98
    globalGemLoadFixAddress = 0x00076BA0
    playBeepAddress = 0x5429c

    # Address of portal surface flags.
    SummerPortalBlock = 0x000e2d34
    AutumnPortalBlock = 0x000f5330
    WinterPortalBlock = 0x000cf28c

    # Guidebook "Has Entered Level" Flags
    CrushGuidebookUnlock = 0x0006B08C
    AutumnGuidebookUnlock = 0x0006B08D
    GulpGuidebookUnlock = 0x0006B098
    WinterGuidebookUnlock = 0x0006B099

    # Level Name Strings
    IdolNameAddress = 0x000106d0
    ColossusNameAddress = 0x000106c4
    HurricosNameAddress = 0x000106b8
    AquariaNameAddress = 0x000106a8
    SunnyNameAddress = 0x0001069c
    OceanNameAddress = 0x0001068c
    SkelosNameAddress = 0x0001065c
    CrystalNameAddress = 0x0001064c
    BreezeNameAddress = 0x0001063c
    ZephyrNameAddress = 0x00066e98
    MetroNameAddress = 0x0001062c
    ScorchNameAddress = 0x00066e90
    ShadyNameAddress = 0x00010620
    MagmaNameAddress = 0x00010614
    FractureNameAddress = 0x00010604
    IcyNameAddress = 0x000105f4
    MysticNameAddress = 0x000105c4
    CloudNameAddress = 0x000105b4
    CanyonNameAddress = 0x000105a4
    RoboticaNameAddress = 0x00010594
    MetropolisNameAddress = 0x00010588
    ShoresNameAddress = 0x00010578

    # Easy Challenge Addresses
    ColossusSpyroHockeyScore = 0x00198ec5
    ColossusOpponentHockeyScore = 0x00198cc9
    IdolFishThrowUp = 0x00082010
    IdolFishIncludeReds = 0x00082048
    IdolFishIncludeRedsHUD = 0x0007efe0
    HurricosLightningThiefStatuses = [
        0x18C8DC,
        0x18CB44,
        0x18CBF4,
        0x18CCA4,
        0x18CCFC,
        0x18CD54,
        0x18CDAC,
        0x18CE04,
        0x18CF0C,
        0x18CFBC
    ]
    HurricosLightningThiefZCoordinates = [
        0x18C8A8,
        0x18CB10,
        0x18CBC0,
        0x18CC70,
        0x18CCC8,
        0x18CD20,
        0x18CD78,
        0x18CDD0,
        0x18CED8,
        0x18CF88
    ]
    spyroHUDScore = 0x00066FEC
    opponentHUDScore = 0x67084
    firstBomboStatus = 0x189f3d
    secondBomboStatus = 0x189b75
    thirdBomboStatus = 0x189805
    bomboAttackAddress = 0x79b34
    fractureHeadbashCheck = 0x78a88
    maxFractureSpiritParticles = 0x64826
    FractureEarthshaperStatuses = [
        0x188DEC,
        0x188E44,
        0x188E9C,
        0x188EF4,
        0x188F4C,
        0x188FA4,
        0x188FFC
    ]
    FractureEarthshaperZCoordinates = [
        0x188DB8,
        0x188E10,
        0x188E68,
        0x188EC0,
        0x188F18,
        0x188F70,
        0x188FC8
    ]
    ShadyHeadbashCheck = 0x7d6c0
    GulpDoubleDamage = 0x120c5e

    # Used to verify game is correct
    GuidebookText = 0x00010308

    # Winter Tundra warp reroute addresses
    WTWallOrbAddress = 0x69fe5
    WTWallOrbBit = 0
    WTDoorGemAddress = 0x6af25
    WTDoorGemBit = 7
    WTWarpAddress = 0x6470a

    lastReceivedArchipelagoID = 0x1c0
    tempLastReceivedArchipelagoID = 0x1c4

class ROM:
    RomDialogueOrbCount = 0x3ec0c
    OrbCountCode = [0x2c, 0x70, 0x22, 0xac]
    RomSFEloraWarp = 0x1057960
    EloraWarpCode = [0x49, 0x7d, 0x00, 0x0c]
    RomAPEloraWarp = 0x24f6a20
    RomLoadOrbCount = 0x152540
    RomSFOrbCount = 0x1050700
    RomAPOrbCount = 0x24f0040
    RomWTOrbCount = 0x405e1a4
    RomPlayBeep = 0x5c9e4
    PlayBeepCode = [0xc8, 0x41, 0x01, 0x0c]
